src/backend/expungeservice/expunger/analyzers/time_analyzer.py

```diff
@@ -58,13 +60,39 @@ def _mark_all_acquittals_ineligible_using_mrd_date(expunger):
         for charge in expunger.acquittals:
             charge.set_time_ineligible('Recommend sequential expungement', eligibility_date)
 
+    @staticmethod
+    def _calculate_has_subsequent_charge(date, charges):
+
+        for charge in charges:
+            if charge.acquitted():
+                date_of_charge = charge.date
+            else:
+                date_of_charge = charge.disposition.date
+
+            if date_of_charge > date:
+                return True
+        return False
+
+
     @staticmethod
     def _evaluate_class_b_felonies(expunger):
-        if expunger.most_recent_charge and expunger.most_recent_charge.disposition.date > TimeAnalyzer.TWENTY_YEARS_AGO:
-            for charge in expunger.class_b_felonies:
-                charge.set_time_ineligible('Time-ineligible under 137.225(5)(a)(A)(i)',
-                                           expunger.most_recent_charge.disposition.date + relativedelta(
-                                               years=TimeAnalyzer.TWENTY_YEARS))
+        for class_b_felony in expunger.class_b_felonies:
+            # If the B felony is acquitted, it is affected by the other time eligibility rules,
+            # so leave it untouched here.
+            # If it's convicted, this restriction overrides any of those rules.
+            if not class_b_felony.acquitted():
+                other_charges = [ charge for charge in expunger.charges if charge != class_b_felony]
+                if TimeAnalyzer._calculate_has_subsequent_charge(class_b_felony.disposition.date, other_charges):
+                    type_eligibility = TypeEligibility(
+                        EligibilityStatus.INELIGIBLE,
+                        reason='137.225(5)(a)(A)(ii) - Class B felony can have no subsequent arrests or convictions')
+                    class_b_felony.expungement_result = ExpungementResult(type_eligibility=type_eligibility, time_eligibility=None)
```

Comment:
This should still be `TimeEligibility`. Maybe we need to rename the classes but my understanding is that `TypeEligibility` is what can be determined with only the charge and `TimeEligibility` is what can be determined only in the context of the entire record.

Reply:
That's true. I was thinking about how the frontend interprets the difference between the two: if a charge is TypeEligible but not TimeEligible, that means it will become eligible at some point in the future, which is not what happens here. This definitely feels a bit off, semantically, though.

Reply:
You could use [`datetime.MAXYEAR`](https://docs.python.org/3/library/datetime.html#datetime.MAXYEAR).

Reply:
Sorry missed this comment earlier. Wouldn't that just result in the frontend displaying, "will be eligible in a billion years"? Ideally it should say Ineligible.

Reply:
Could do something like having time eligibiltiy `NEVER`. I agree though that we should keep the TypeEligibility and reason intact to stay clear to the user.

Reply:
A slightly long but accurate name for the two fields are "Default type eligibility" and "Eligibility in context of record and date". For Class B Felonies with subsequent arrests, we can use the value of `None` for the `date_will_be_eligible`. Records that are eligible in the context of record and date should have the date they became eligible in the `date_will_be_eligible` instead of the current value of `None`. This way `None` always means never in this context, which I think is an acceptable use of this often abused value.

We should then have a separate serialized field for the largest display label that we can build up right before the serializer using the DefaultEligibility/EligibilityInRecordContext (TypeEligibility/TimeEligibility). [We were looking to push this logic from the frontend to the backend anyways](https://github.com/codeforpdx/recordexpungPDX/pull/557#issuecomment-564383275).

I think it is fine to do this in a separate PR.

Reply:
A class b felony can be type eligible; however, if there is a subsequent charge then it invalidates the eligibility forever. This is definitely out of place and confusing here. Have we asked Michael how he would like to display this scenario?

Reply:
The main part that matters for the displayed result is: The top line on the charge is the red "Ineligible" label, and the displayed reason whether it's from the Type eligibilty for the Time eligibility, is clear that it's ineligible because there was a subsequent arrest or conviction. As Kent said, TypeEligibility is a function of the charge analyzed by itself whereas time eligibility is a function of the record as a whole, and I'm inclined to stick with that. The way we render the result can be a function of that underlying specification when we present to the user.